### PR TITLE
Revert change to use `BinaryFormatter` for `Out-GridView`

### LINKED ISSUE #20223: Out-GridView filtering via criteria in 7.4.0-preview.5 yields incorrect results
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Using the "Add criteria" and filtering that way in 7.4.0-preview.5 yields incorrect results:

```powershell
@(@{"a" = '=='},@{"b"='<='},@{"c"='>='}) |ogv
```

image

Here's the correct outcome with 7.3.6:

image

### Expected behavior

```console
Objects matching the specified filter should be displayed.
```

### Actual behavior

```console
Filtering removes all objects even those matching the criteria.
```

### Error details

_No response_

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.0-preview.5
PSEdition                      Core
GitCommitId                    7.4.0-preview.5
OS                             Microsoft Windows 10.0.22621
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

Revert  #20050

`Out-GridView` was designed around using the `BinaryFormatter` to make deep copies of Filters.  When we removed usage of that type, we inadvertently broke filtering (although `contains` filter still worked, but only that one).  This change reverts the specific removal of use of `BinaryFormatter` for `Out-GridView` to bring back the previous functionality.  The previous change to a `Clone()` method was removed as it was incomplete.

Manually tested each filter and duplicate filters.

## PR Context

Fix https://github.com/PowerShell/PowerShell/issues/20223

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by iSazonov** [comment]
BinaryFormatter will removed from .Net 9.0 at all. So maybe revert only for release branch 7.4 and implement full fix for master branch?

**Comment by SteveL-MSFT** [comment]
@iSazonov yes, we have more time with 7.5 to change the design to not rely on BinaryFormatter

File: src/Microsoft.Management.UI.Internal/ManagementList/FilterCore/FilterRules/FilterRuleExtensions.cs [hunk 1]
```diff
@@ -27,7 +27,34 @@ public static class FilterRuleExtensions
...
         public static FilterRule DeepCopy(this FilterRule rule)
         {
-            return rule.Clone();
+            ArgumentNullException.ThrowIfNull(rule);
+
+#pragma warning disable SYSLIB0050
+            Debug.Assert(rule.GetType().IsSerializable, "rule is serializable");
+
+#pragma warning disable SYSLIB0011
+#pragma warning disable SYSLIB0050
```
Comment (daxian-dbw) [inline src/Microsoft.Management.UI.Internal/ManagementList/FilterCore/FilterRules/FilterRuleExtensions.cs:36]: Those warnings are repeatedly disabled, which seems to be unnecessary. Can you maybe just disable both warnings for this whole method and restore them at the end of the method?

**Review by daxian-dbw — COMMENTED** [review]
Looks good to me overall. Left 1 comment.

File: src/Microsoft.Management.UI.Internal/ManagementList/FilterCore/FilterRules/FilterRuleExtensions.cs (hunk 1 above)
Comment (SteveL-MSFT) [reply src/Microsoft.Management.UI.Internal/ManagementList/FilterCore/FilterRules/FilterRuleExtensions.cs]: You're right, new commit should do this optimally

**Review by daxian-dbw — APPROVED** [review]

**Comment by iSazonov** [comment]
Have we a tracking issue for follow work?

**Pull request merged**